pkg/engine/internal/executor/pipeline.go

```diff
@@ -248,3 +285,62 @@ func (lp *lazyPipeline) Close() {
 	}
 	lp.built = nil
 }
+
+// Region implements RegionProvider.
+func (lp *lazyPipeline) Region() *xcap.Region {
+	if lp.built != nil {
+		if provider, ok := lp.built.(RegionProvider); ok {
+			return provider.Region()
+		}
+	}
+	return nil
+}
+
+// observedPipeline wraps a Pipeline to automatically collect common statistics
+// and record them to the pipeline's region.
+type observedPipeline struct {
+	inner  Pipeline
+	region *xcap.Region
+}
+
+var _ Pipeline = (*observedPipeline)(nil)
+
+// newObservedPipeline wraps a pipeline to automatically collect common statistics.
+// If the pipeline has a region, statistics will be recorded to it.
+func newObservedPipeline(inner Pipeline) *observedPipeline {
```

Comment:
This is only called when executing ScanSet, so we'd only get these stats for the basic engine where the scheduler isn't being used; is that intentional?

Reply:
updated in https://github.com/grafana/loki/pull/19821/commits/d21f01b71d6f0907e8adef86221a35f829fcfd4d